File: app/src/main/java/club/ranleng/psnine/base/BaseFragment.java
package club.ranleng.psnine.base;

import android.app.Fragment;
import android.content.Context;
import android.os.Bundle;
import android.support.annotation.Nullable;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import com.blankj.utilcode.util.NetworkUtils;
import com.blankj.utilcode.util.ToastUtils;

import club.ranleng.psnine.R;
import club.ranleng.psnine.common.Key;
import club.ranleng.psnine.ui.main.MainActivity;

public abstract class BaseFragment extends Fragment {

    public Context mContext;
    protected View mRootView;
    protected boolean isVisible;
    private boolean isPrepared;
    private boolean isLoadData = false;

    @Override
    public void setUserVisibleHint(boolean isVisibleToUser) {
        super.setUserVisibleHint(isVisibleToUser);
        if (getUserVisibleHint()) {
            isVisible = true;
            lazyLoad();
        } else {
            isVisible = false;
            onInvisible();
        }
    }


    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        mContext = getActivity();
        setHasOptionsMenu(true);
    }


    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {
        if (mRootView == null) {
            mRootView = initView(inflater, container);
        }
        return mRootView;
    }

    @Override
    public void onActivityCreated(@Nullable Bundle savedInstanceState) {
        super.onActivityCreated(savedInstanceState);
        isPrepared = true;
        lazyLoad();
    }

    /**
     * 懒加载
     */
    protected void lazyLoad() {
        if (!NetworkUtils.isConnected()) {
            ToastUtils.showShort(R.string.no_internet);
            return;
        }

        if (!isPrepared || isLoadData) {
            return;
        }

        if (Key.getSetting().PREF_PRELOAD || !(getActivity() instanceof MainActivity)) {
            isLoadData = true;
            initData();
            return;
        }

        if (!isVisible) {
            return;
        }

        isLoadData = true;
        initData();

    }

    // 不可见
    protected void onInvisible() {

    }

    public abstract View initView(LayoutInflater inflater, ViewGroup container);

    public abstract void initData();

}

File: app/src/main/java/xyz/rankki/psnine/base/BaseFragment.kt
package xyz.rankki.psnine.base

import android.os.Bundle
import android.support.v4.app.Fragment
import android.support.v4.app.FragmentActivity
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup

abstract class BaseFragment : Fragment() {


    lateinit var mContext: FragmentActivity
    private var mRootView: View? = null
    private var isPrepared: Boolean = false
    private var isLoadData: Boolean = false
    private var visibility: Boolean = false

    override fun setUserVisibleHint(isVisibleToUser: Boolean) {
        super.setUserVisibleHint(isVisibleToUser)
        this.visibility = this.userVisibleHint
        if (this.visibility) {
            lazyLoad()
        }
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        this.mContext = activity!!
        setHasOptionsMenu(true)
    }

    override fun onCreateView(inflater: LayoutInflater,
                              container: ViewGroup?,
                              savedInstanceState: Bundle?): View? {
        if (mRootView == null) {
            mRootView = initView()
        }
        return mRootView
    }

    override fun onActivityCreated(savedInstanceState: Bundle?) {
        super.onActivityCreated(savedInstanceState)
        this.isPrepared = true
        lazyLoad()
    }

    private fun lazyLoad() {

        if (!isPrepared || isLoadData) {
            return
        }

        if (!visibility) {
            return
        }

        initData()
        this.isLoadData = true

    }

    abstract fun initView(): View

    abstract fun initData()
}